modules/globebrowsing/util/globelabelscomponent.cpp

```diff
@@ -0,0 +1,842 @@
+
+        bool success = fileStream.good();
+        return success;
+    }
+    else {
+        LERROR(fmt::format("Error opening file '{}' for loading cache file", file));
+        return false;
+    }
+}
+
+bool GlobeLabelsComponent::saveCachedFile(const std::string& file) const {
+
+    std::ofstream fileStream(file, std::ofstream::binary);
+    if (fileStream.good()) {
+        fileStream.write(reinterpret_cast<const char*>(&CurrentCacheVersion),
+            sizeof(int8_t));
+
+        int32_t nValues = static_cast<int32_t>(_labels.labelsArray.size());
+        if (nValues == 0) {
+            LERROR("Error writing cache: No values were loaded");
+            return false;
+        }
+        fileStream.write(reinterpret_cast<const char*>(&nValues), sizeof(int32_t));
+
+        size_t nBytes = nValues * sizeof(_labels.labelsArray[0]);
+        fileStream.write(reinterpret_cast<const char*>(&_labels.labelsArray[0]), nBytes);
+
+        bool success = fileStream.good();
+        return success;
+    }
+    else {
+        LERROR(fmt::format("Error opening file '{}' for save cache file", file));
+        return false;
+    }
+}
+
+void GlobeLabelsComponent::draw(const RenderData& data) {
+    if (_labelsEnabled) {
+        // Calculate the MVP matrix
+        glm::dmat4 viewTransform = glm::dmat4(data.camera.combinedViewMatrix());
+        glm::dmat4 vp = glm::dmat4(data.camera.sgctInternal.projectionMatrix()) *
+            viewTransform;
+        glm::dmat4 mvp = vp * _globe->modelTransform();
+
+        glm::dvec3 globePositionWorld = glm::dvec3(_globe->modelTransform() * 
+            glm::vec4(0.0, 0.0, 0.0, 1.0));
+        glm::dvec3 cameraToGlobeDistanceWorld = globePositionWorld - 
+            data.camera.positionVec3();
+        double distanceCameraGlobeWorld = glm::length(cameraToGlobeDistanceWorld);
+
+        float fadeInVariable = 1.f;
+        if (_labelsFadeInEnabled) {
+            double averageRadius = (
+                _globe->ellipsoid().radii().x + _globe->ellipsoid().radii().y + 
+                _globe->ellipsoid().radii().z
+                ) / 3.0;
+            glm::dvec2 fadeRange = glm::dvec2(
+                averageRadius + _labelsMinHeight
+            );
+            fadeRange.x += _labelsFadeInDist;
+            double a = 1.0 / (fadeRange.y - fadeRange.x);
+            double b = -(fadeRange.x / (fadeRange.y - fadeRange.x));
+            double funcValue = a * distanceCameraGlobeWorld + b;
+            fadeInVariable *= funcValue > 1.0 ? 1.f : static_cast<float>(funcValue);
+
+            if (fadeInVariable < 0.009f) {
+                return;
+            }
+        }
+
+        if (_labelsFadeOutEnabled) {
+            double averageRadius = (
+                _globe->ellipsoid().radii().x + _globe->ellipsoid().radii().y +
+                _globe->ellipsoid().radii().z
+                ) / 3.0;
+            glm::dvec2 fadeRange = glm::dvec2(
+                averageRadius + _labelsMinHeight + 1500.0
```

Comment:
There are some "magic" constants here. Consider moving to anonymous namespace, or at least add comments that motivate them.